### PR TITLE
Add temporal support to QField

### PR DESCRIPTION
**Pull request by nirvn**
This PR adds temporal awareness and filtering to QField :partying_face: 

First, a screencast

https://user-images.githubusercontent.com/1728657/154828334-bec77f6f-6d93-433d-aa2e-cc62132926f5.mp4

A few implementation details:
- The temporal filter button in the side dashboard will only appear if a given project has temporal layer(s) already set up
- The temporal filter settings are remembered across sessions
- The temporal filter settings _will_ narrow down canvas tap feature identification (very useful when dealing with large datasets that can crowd canvas view when temporally unfiltered)
- A calendar panel popup widget has been created (essentially taking relevant bits from the editor datetime widget), a subsequent PR will remove duplicate code from the editor datetime widget

This was coded as a gift to a good friend and collaborator of mine, hope the broader community will enjoy too :)

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 20e476cb9a3187ad871b921a85251a166215303a:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-20e05f42a81459a0f3605aaccef753e502b912be-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-20e05f42a81459a0f3605aaccef753e502b912be-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-20e05f42a81459a0f3605aaccef753e502b912be-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-20e05f42a81459a0f3605aaccef753e502b912be-x86-android.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for dc2823a4358f1cc474df151aa12ca9f940d6b7a5:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-9ec0ef44a5f340dc70430fc4e67f451483892080-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-9ec0ef44a5f340dc70430fc4e67f451483892080-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-9ec0ef44a5f340dc70430fc4e67f451483892080-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-9ec0ef44a5f340dc70430fc4e67f451483892080-x86-android.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 759b3ff6b7506f57f9325d6be5a07ce3f329b45e:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-4cc09753d25243f98e7c77106e2685fb5273fe8c-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-4cc09753d25243f98e7c77106e2685fb5273fe8c-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-4cc09753d25243f98e7c77106e2685fb5273fe8c-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-4cc09753d25243f98e7c77106e2685fb5273fe8c-x86-android.apk)

File: src/core/layertreemodel.cpp [hunk 1]
```diff
@@ -959,6 +966,21 @@ void FlatLayerTreeModelBase::updateCurrentMapTheme()
   }
 }
 
+void FlatLayerTreeModelBase::updateTemporalState()
+{
+  mIsTemporal = false;
+  QVector<QgsMapLayer *> mapLayers = QgsProject::instance()->layers<QgsMapLayer *>();
```
Comment (m-kuhn) [inline src/core/layertreemodel.cpp:972]:
```suggestion
  QVector<QgsMapLayer *> mapLayers = mProject->layers<QgsMapLayer *>();
```

**Review by m-kuhn — APPROVED** [review]
🎉

**Comment by Jimzi** [comment]
Hi Mathieu,

I have just started converting my project to make use of temporal features of QGIS and this is a great addition to QField. My sincerest thanks for sharing your friend's gift. 👍

**Comment by nirvn** [comment]
@Jimzi , thanks for dropping by and sharing your appreciation of QField. Out of pure curiosity, what kind of data & project are you making temporarily aware?

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 21f245f0783f982494fa47fcf8eb280c12169524:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-bdd583ea57ecdfc34275b7378c82888a7fe3cad1-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-bdd583ea57ecdfc34275b7378c82888a7fe3cad1-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-bdd583ea57ecdfc34275b7378c82888a7fe3cad1-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2528-bdd583ea57ecdfc34275b7378c82888a7fe3cad1-x86-android.apk)

**Comment by Jimzi** [comment]
@nirvn , I am keeping track of trees on our farm, palm trees to be specific.  Part of the reason is tracking diseases that affect certain trees and trying to see if there is any relation between their location, infection durations, treatments and so on.

Sometimes these trees are rooted out when untreatable and replaced with new ones, but I would like to maintain the history of what trees were where, when they were infected and how they were treated, instead of just losing that data once the trees were gone.

I had initially created an _'archived'_ field to filter out nonexistent features from showing while still being held in the geopackage. Then I added an _'archived date'_ and _'archived comment'_ for flexibility and my queries started getting bigger for even the "non temporal?" queries, and that's when I started searching for existing solutions and discovered that there are temporal databases, and was delighted when I discovered QGIS supported such implementations. I am still trying to figure it out and just beginning to discover how these things can be implemented. 

It is no where near the advanced uses I've seen online while searching temporal implementation (e.g. meteorology), so I'm hoping it not too hard for a simple  farmer trying non-traditional new tools :)

**Pull request merged**